### PR TITLE
feat(protocol,supplementary-contracts): relocate & allow TokenUnlock to deploy and own ProverSets

### PR DESCRIPTION
**Pull request by dantaik**
Allow Taiko's investors to use their locked tokens as bonds.

### CODE REVIEW

**Comment by Brechtpd** [comment]
Wouldn't this allow people to withdraw a big chunk of their tokens by getting slashed and then getting the reward?

 I guess would have to think about the specifics and some protocol details, but for example if you can prove a block wrongly and directly also contest and prove it correctly in the same tx you would be able to extract a big chunk without any risks of anybody else getting the tokens. So sure you still lose a big chunk, but you also don't have to wait 4 years to get your tokens.

**Comment by dantaik** [comment]
> Wouldn't this allow people to withdraw a big chunk of their tokens by getting slashed and then getting the reward?
> 
> I guess would have to think about the specifics and some protocol details, but for example if you can prove a block wrongly and directly also contest and prove it correctly in the same tx you would be able to extract a big chunk without any risks of anybody else getting the tokens. So sure you still lose a big chunk, but you also don't have to wait 4 years to get your tokens.

Good points. I believe technically it is possible. However, I think practically this won't happen for two reasons:

- 12.5% token will be lost during the contest-then-re-prove process.
- all of our blocks requires SGX, they must be able to generate incorrect but verifiable sgx proofs, then overwrite them with valid ones.

**Comment by Brechtpd** [comment]
Yeah true that if SGX is required for each block it shouldn't be possible, so should be good.

**Review by Brechtpd — APPROVED** [review]

**Review by davidtaikocha — APPROVED** [review]

**Comment by Brechtpd** [comment]
I'm not 100% sure about the behavior for this in the protocol right now, but the liveness bond slashing could also be a potential way to avoid the locking. But that one is probably a lot trickier to pull off risk free.

**Comment by dantaik** [comment]
> liveness

As of now liveness bond is the smallest bond compared with all other bonds. When liveness bond is returned, another bond must be placed.

File: packages/protocol/contracts/team/proving/ProverSet.sol [hunk 1]
```diff
@@ -1,63 +1,92 @@
...
 
         emit ProverEnabled(_prover, _isProver);
     }
 
-    /// @notice Withdraws Taiko tokens back to the owner address.
-    function withdraw(uint256 _amount) external onlyOwner {
-        IERC20(resolve(LibStrings.B_TAIKO_TOKEN, false)).transfer(owner(), _amount);
+    /// @notice Withdraws Taiko tokens back to the admin address.
+    function withdrawToAdmin(uint256 _amount) external onlyAuthorized {
+        IERC20(resolve(LibStrings.B_TAIKO_TOKEN, false)).transfer(admin, _amount);
```
Comment (adaki2004) [inline packages/protocol/contracts/team/proving/ProverSet.sol:75]: I guess we need to put this function in the `TokenUnlock` too, if we want to allow unlocking people calling it ?

File: packages/protocol/contracts/team/proving/ProverSet.sol [hunk 2]
```diff
@@ -1,63 +1,92 @@
...
-        address assignmentHook = resolve(LibStrings.B_ASSIGNMENT_HOOK, false);
-        tko.approve(assignmentHook, type(uint256).max);
+        IERC20 tko = IERC20(resolve(LibStrings.B_TAIKO_TOKEN, false));
+        tko.approve(resolve(LibStrings.B_TAIKO, false), type(uint256).max);
+        tko.approve(resolve(LibStrings.B_ASSIGNMENT_HOOK, false), type(uint256).max);
     }
 
     /// @notice Enables or disables a prover.
-    function enableProver(address _prover, bool _isProver) external onlyOwner {
+    function enableProver(address _prover, bool _isProver) external onlyAuthorized {
```
Comment (adaki2004) [inline packages/protocol/contracts/team/proving/ProverSet.sol:66]: I guess we need to put this function in the `TokenUnlock` too, if we want to allow unlocking people calling it.

File: packages/protocol/contracts/team/proving/ProverSet.sol [hunk 3]
```diff
@@ -1,63 +1,92 @@
...
+    function proveBlock(uint64 _blockId, bytes calldata _input) external onlyProver nonReentrant {
         emit BlockProvenBy(msg.sender, _blockId);
         ITaikoL1(resolve(LibStrings.B_TAIKO, false)).proveBlock(_blockId, _input);
     }
 
+    /// @notice Delegates token voting right to a delegatee.
+    /// @param _delegatee The delegatee to receive the voting right.
+    function delegate(address _delegatee) external onlyAuthorized nonReentrant {
+        ERC20VotesUpgradeable(resolve(LibStrings.B_TAIKO_TOKEN, false)).delegate(_delegatee);
+    }
```
Comment (adaki2004) [inline packages/protocol/contracts/team/proving/ProverSet.sol:88]: This one too, shall be put also to `TokenUnlocking` (to allow calling into this function) ?

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 1 above)
Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol:75]: No need, the `onlyAuthorized` allows recipient to call this function.

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 2 above)
Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol:66]: why the unlocking needs to know about prover enabling...

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 3 above)
Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol:88]: Sorry I don't understand this feedback, why duplicate logics in two contracts?

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 2 above)
Comment (adaki2004) [reply packages/protocol/contracts/team/proving/ProverSet.sol:66]: Yes, okay i see the modifier is like that:
`msg.sender != IHasRecipient(admin).recipient()`

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 1 above)
Comment (adaki2004) [reply packages/protocol/contracts/team/proving/ProverSet.sol:75]: Yes, okay i see the modifier is like that:
`msg.sender != IHasRecipient(admin).recipient()`

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 3 above)
Comment (adaki2004) [reply packages/protocol/contracts/team/proving/ProverSet.sol:88]: Yep, i missed the modifier change. Seems no issue indeed.

**Review by adaki2004 — APPROVED** [review]

**Pull request merged**